File: study/ts/script/manager/GatewayListManager.ts
import { config } from "../config";
import { GatewayData } from "../data/GatawayData";
import { G_ConfigManager } from "../init";
import { PrioritySignal } from "../utils/event/PrioritySignal";
import { handler } from "../utils/handler";
import ALDStatistics from "../utils/ALDStatistics";

export class GatewayListManager {
    private _list: GatewayData[];
    private _inited;
    private _useTestList;
    private _lastRemoteTime;
    private _lastRequestTime;
    private _lastRandServer;
    private _httpRequest: XMLHttpRequest;
    public signal;

    constructor() {
        this._inited = false;
        this._useTestList = false;
        this._lastRemoteTime = 0;
        this._lastRequestTime = 0;
        this._lastRandServer = null;
        this._list = [];
        this.signal = new PrioritySignal('string');
    }

    public getGateway(): GatewayData {
        return this._list[0];
    }

    public isCheckUpdate() {
        if (CC_DEBUG) return true;
        var time = G_ConfigManager.getGatewayCacheTime() || 900;
        if (this._lastRemoteTime == 0 || new Date().getTime() / 1000 - this._lastRemoteTime > time || this._list.length == 0) {
            return true;
        }
        return false;
    }

    public checkUpdateList() {
        //QA测试：
        if (config.ENV < 3) {
            this._list = [];
            let gatewayData = new GatewayData();
            gatewayData.setIp(config.WEBSOCKET_IP);
            gatewayData.setPort(config.WEBSOCKET_PORT);
            this._list.push(gatewayData)
            this.signal.dispatch('success');
            return;
        } else {
            if (this.isCheckUpdate()) {
                var remoteGateway = G_ConfigManager.isRemoteGateway();
                if (remoteGateway) {
                    this._getRemoteServerList();
                } else {
                    var ret = G_ConfigManager.getListGateway();
                    if (ret != null && ret != '') {
                        var infos = this._decode(ret);
                        if (infos.length > 0) {
                            var list = [];
                            for (let i = 0; i < infos.length; i++) {
                                var info = infos[i];
                                var gateway = new GatewayData(info);
                                list.push(gateway);
                            }
                            this.setGatewayList(list);
                        }
                    }
                    this.signal.dispatch('success');
                }
            } else {
                this.signal.dispatch('success');
            }
        }
    }
    public setGatewayList(list) {
        this._list = list;
    }
    private _onCheckGatewayCache() {
        this._getRemoteServerList();
    }
    private _getRemoteServerList() {
        if (this._httpRequest == null) {
            var url = config.GATEWAYLIST_URL_TEMPLATE;
            url = url.replace('#domain#', config.GATEWAYLIST_URL);
            this._httpRequest = new XMLHttpRequest();
            // this._httpRequest.responseType =XMLHTTPREQUEST_RESPONSE_STRING;
            this._httpRequest.open('GET', url);
            this._httpRequest.onreadystatechange = (handler(this, this._onReadyStateChange));
            this._httpRequest.onerror = this._httpRequest.ontimeout = () => {
                ALDStatistics.instance.aldSendEvent('获取服务器地址和端口失败');
                this.signal.dispatch('fail');
            };
            this._httpRequest.send();
        }
    }
    private _onReadyStateChange() {
        var e = 'fail';
        if (this._httpRequest.readyState == 4) {
            if (this._httpRequest.status >= 200 && this._httpRequest.status < 207) {
                var response = this._httpRequest.response;
                var infos = this._decode(this._httpRequest.response);
                if (infos.length > 0) {
                    var list = [];
                    var addServerList = G_ConfigManager.getAddGateway();
                    if (addServerList != null && addServerList != '') {
                        var list = this._decode(addServerList);
                        for (let k in list) {
                            var v = list[k];
                            var gateway = new GatewayData(v);
                            list.push(gateway);
                        }
                    }
                    for (let i = 0; i < infos.length; i++) {
                        var info = infos[i];
                        var gateway = new GatewayData(info);
                        list.push(gateway)
                    }
                    this.setGatewayList(list);
                    this._lastRemoteTime = new Date().getTime() / 1000;
                    e = 'success';
                }
            }
            this._httpRequest.abort();
            this._httpRequest = null;
            this.signal.dispatch(e);
        }
    }
    private _decode(response: string): any[] {
        var cache = [];
        var gateways: string[] = response.split(',');
        if (gateways && gateways.length >= 1) {
            for (let i in gateways) {
                var v = gateways[i];
                var ret = v.split('|');
                var info: any = {};
                info.ip = ret[0];
                if (info.ip.indexOf('wss://') == -1) {
                    info.ip = 'wss://' + info.ip;
                }
                info.port = 0;
                if (ret[1] != null) {
                    info.port = parseInt(ret[1]);
                }
                cache.push(info);
            }
        }
        return cache;
    }
}

File: study/lua/src/app/manager/GatewayListManager.lua
local GatewayListManager = class("GatewayListManager")
local scheduler = require("cocos.framework.scheduler")
local PrioritySignal = require("yoka.event.PrioritySignal")
local GatewayData = require("app.data.GatewayData")

--
function GatewayListManager:ctor()
	self._inited = false
    self._useTestList = false
    self._lastRemoteTime = 0
    self._lastRequestTime = 0
    self._lastRandServer = nil
    self._list = {}
    self.signal = PrioritySignal.new("string")
end

--
function GatewayListManager:clear()
    
end

--
function GatewayListManager:reset()
    
end

--
function GatewayListManager:getGateway()
    return self._list[1]
end

--
function GatewayListManager:isCheckUpdate()
    local time = G_ConfigManager:getGatewayCacheTime() or 900
    if self._lastRemoteTime == 0 or (timer:gets() - self._lastRemoteTime) > time or #self._list == 0 then
        return true
    end
    
    return false
end

--
function GatewayListManager:checkUpdateList()
    if self:isCheckUpdate() then
        local remoteGateway = G_ConfigManager:isRemoteGateway()
        if remoteGateway then
            --[[if self._timer == nil then
                self._timer = scheduler.scheduleGlobal(handler(self, self._onCheckGatewayCache), 900)
            end]]
            self:_getRemoteServerList()
        else
            local ret = G_ConfigManager:getListGateway()
            if ret ~= nil and ret ~= "" then
                local infos = self:_decode(ret)
                if #infos > 0 then
                    local list = {}
                    for i=1, #infos do
                        local info = infos[i]
                        local gateway = GatewayData.new(info)
                        list[#list + 1] = gateway
                    end
                    self:setGatewayList(list)
                end
            end

            self.signal:dispatch("success")
        end
    else
        self.signal:dispatch("success")
    end
end

--
function GatewayListManager:setGatewayList(list)
	self._list = list
end

--
function GatewayListManager:_onCheckGatewayCache()
    self:_getRemoteServerList()
end

--
function GatewayListManager:_getRemoteServerList()
    if self._httpRequest == nil then
        print("GatewayListManager:_getRemoteServerList")
        local url = GATEWAYLIST_URL_TEMPLATE
        url = string.gsub(url, "#domain#", GATEWAYLIST_URL)
        self._httpRequest = cc.XMLHttpRequest:new()
        self._httpRequest.responseType = cc.XMLHTTPREQUEST_RESPONSE_STRING
        self._httpRequest:open("GET", url)
        self._httpRequest:registerScriptHandler(handler(self, self._onReadyStateChange))
        self._httpRequest:send()
    end
end

--
function GatewayListManager:_onReadyStateChange()
    local e = "fail"
    if self._httpRequest.readyState == 4 and (self._httpRequest.status >= 200 and self._httpRequest.status < 207) then
        local response = self._httpRequest.response
        local infos = self:_decode(self._httpRequest.response)
        if #infos > 0 then
            dump(infos)
            local list = {}
            local addServerList = G_ConfigManager:getAddGateway()
            if addServerList ~= nil and addServerList ~= "" then
                local list = self:_decode(addServerList)
                for k,v in ipairs(list) do
                    local gateway = GatewayData.new(v)
                    list[#list + 1] = gateway
                end
            end

            for i=1, #infos do
                local info = infos[i]
                local gateway = GatewayData.new(info)
                list[#list + 1] = gateway
            end
            self:setGatewayList(list)
            self._lastRemoteTime = timer:gets()
            e = "success"
        end
    end

    self._httpRequest:unregisterScriptHandler()
    self._httpRequest = nil
    self.signal:dispatch(e)
end

--
function GatewayListManager:_decode(response)
    local cache = {}
    local gateways = string.split(response, ",")
    if gateways and #gateways >= 1 then
        for i,v in ipairs(gateways) do
            local ret = string.split(v, "|")
            local info = {}
            info.ip = ret[1]
            info.port = 0
            if ret[2] ~= nil then
                info.port = tonumber(ret[2])
            end
            cache[#cache + 1] = info
        end
    end

    return cache
end

return GatewayListManager